File: plugins/DEV-ESSENTIALS/hooks/suggest-compact.js
#!/usr/bin/env node
/**
 * PreToolUse Hook: Suggest Compact
 *
 * Counts tool invocations and suggests /compact every ~50 calls
 * to manage context window size.
 *
 * Non-blocking (exit code 0).
 */

const fs = require('fs');
const path = require('path');
const os = require('os');
const MAX_STDIN = 1024 * 1024;

const COMPACT_INTERVAL = 50;
const STATE_FILE = path.join(os.tmpdir(), '.claude-hook-tool-count');

let data = '';
process.stdin.setEncoding('utf8');

process.stdin.on('data', chunk => {
  if (data.length < MAX_STDIN) data += chunk.substring(0, MAX_STDIN - data.length);
});

process.stdin.on('end', () => {
  try {
    const disabled = (process.env.DISABLED_HOOKS || '').split(',');
    if (disabled.includes('suggest-compact')) {
      process.stdout.write(data);
      return;
    }

    let count = 0;
    try {
      count = parseInt(fs.readFileSync(STATE_FILE, 'utf8').trim(), 10) || 0;
    } catch {
      // File doesn't exist yet
    }

    count++;
    fs.writeFileSync(STATE_FILE, String(count), 'utf8');

    if (count > 0 && count % COMPACT_INTERVAL === 0) {
      console.error(`[Hook] You've made ~${count} tool calls this session. Consider running /compact to manage context window size.`);
    }

    process.stdout.write(data);
  } catch (err) {
    console.error(`[Hook] suggest-compact error: ${err.message}`);
    process.stdout.write(data);
  }
});

File: plugins/DEV-ESSENTIALS/hooks/suggest-compact.py
#!/usr/bin/env python3
"""
PreToolUse Hook: Suggest Compact

Counts tool invocations and suggests /compact every ~50 calls
to manage context window size.

Non-blocking (exit code 0).
"""

import json
import os
import sys
import tempfile

MAX_STDIN = 1024 * 1024
COMPACT_INTERVAL = 50
STATE_FILE = os.path.join(tempfile.gettempdir(), ".claude-hook-tool-count")


def main():
    data = sys.stdin.read(MAX_STDIN)

    try:
        disabled = os.environ.get("DISABLED_HOOKS", "").split(",")
        if "suggest-compact" in disabled:
            sys.stdout.write(data)
            return

        count = 0
        try:
            with open(STATE_FILE, "r") as f:
                count = int(f.read().strip())
        except (FileNotFoundError, ValueError):
            pass

        count += 1

        with open(STATE_FILE, "w") as f:
            f.write(str(count))

        if count > 0 and count % COMPACT_INTERVAL == 0:
            print(f"[Hook] You've made ~{count} tool calls this session. Consider running /compact to manage context window size.", file=sys.stderr)

        sys.stdout.write(data)
    except Exception as err:
        print(f"[Hook] suggest-compact error: {err}", file=sys.stderr)
        sys.stdout.write(data)


if __name__ == "__main__":
    main()
